dagshub/data_engine/model/query.py

```diff
@@ -183,31 +191,47 @@ def _serialize_node(node: Node, tree: Tree) -> Dict:
             value = node.data["value"]
             as_of = node.data.get("as_of")
 
-            value_type = metadataTypeLookup[type(value)].value
-            if type(value) is bytes:
-                # TODO: this will need to probably be changed when we allow actual binary field comparisons
-                value = value.decode("utf-8")
+            if query_op in [FieldFilterOperand.YEAR,
+                            FieldFilterOperand.MONTH,
+                            FieldFilterOperand.DAY,
+                            FieldFilterOperand.TIMEOFDAY]:
+                res = {
+                    "filter": {
+                        "key": key,
+                        "value": value if query_op is FieldFilterOperand.TIMEOFDAY else 0,
+                        "valueRange": 0 if query_op is FieldFilterOperand.TIMEOFDAY else value,
+                        "valueType": "DATETIME_RANGE",
+                        "comparator": query_op.value,
+                    }
+                }
             else:
-                if isinstance(value, datetime.datetime):
-                    value = int(value.timestamp())
+                value_type = metadataTypeLookup[type(value)].value
+                if type(value) is bytes:
+                    # TODO: this will need to probably be changed when we allow actual binary field comparisons
+                    value = value.decode("utf-8")
                 else:
-                    value = str(value)
-
-            if value_type is None:
-                raise RuntimeError(
-                    f"Value type {value_type} is not supported for querying.\r\n"
-                    f"Supported types: {list(metadataTypeLookup.keys())}"
-                )
-            res = {
-                "filter": {
-                    "key": key,
-                    "value": str(value),
-                    "valueType": value_type,
-                    "comparator": query_op.value,
+                    if isinstance(value, datetime.datetime):
+                        value = int(value.timestamp() * 1000)
+                    else:
+                        value = str(value)
+
+                if value_type is None:
+                    raise RuntimeError(
+                        f"Value type {value_type} is not supported for querying.\r\n"
+                        f"Supported types: {list(metadataTypeLookup.keys())}"
+                    )
+                res = {
+                    "filter": {
+                        "key": key,
+                        "value": str(value),
+                        "valueType": value_type,
+                        "comparator": query_op.value,
+                    }
                 }
-            }
```

Comment:
I think having an if/else with different constructors for the `res` will lead to bugs later where someone accidentally changes one but not the other.
I initially wanted to suggest changing the res.value fiteld after in an if, but then realized that you probably hit a problem with the `value_type is None` check.

So something like this:
```python
dt_range_ops = [FieldFilterOperand.YEAR,
                            FieldFilterOperand.MONTH,
                            FieldFilterOperand.DAY,
                            FieldFilterOperand.TIMEOFDAY]
value_type = metadataTypeLookup[type(value)].value
if type(value) is bytes:
    # TODO: this will need to probably be changed when we allow actual binary field comparisons
    value = value.decode("utf-8")
else:
    if isinstance(value, datetime.datetime):
        value = int(value.timestamp())
    else:
        value = str(value)

if value_type is None and query_op not in dt_range_ops:
    raise RuntimeError(
        f"Value type {value_type} is not supported for querying.\r\n"
        f"Supported types: {list(metadataTypeLookup.keys())}"
    )
res = {
    "filter": {
        "key": key,
        "value": str(value),
        "valueType": value_type,
        "comparator": query_op.value,
    }
}
if query_op in dt_range_ops:
  res["value"] = ...
  res["valueType"] = ...
```